### PR TITLE
fix: implement graceful shutdown for api server

### LINKED ISSUE #18642: Implement graceful shutdown in all Argo CD components
Procfile sets GOCOVDIR for five Argo CD components. But only one (applicationset-controller) shuts down gracefully enough for code coverage data to be produced. In other words, go writes a covmeta file, but not a covcounters file (the file that actually contains coverage info). So today we get partial code coverage info for e2e tests.

The task is to do two things for the remaining four components: 1) implement graceful shutdown and 2) [add the component's gocov directory to the e2e reporting CI](https://github.com/argoproj/argo-cd/blob/master/.github/workflows/ci-build.yaml#L367-L371).

- [x] applicationset-controller
- [x] controller
- [ ] api-server ([WIP](https://github.com/argoproj/argo-cd/pull/15938))
- [x] repo-server
- [x] notification

Exactly what "graceful shutdown means" may vary by component. Concretely, I think it means we need to listen to shutdown signals and end all running goroutines. But the "graceful" part depends on what's running.

### PR DESCRIPTION
**Pull request by leoluz**
Argo CD API server will eventually return 502 during rolling update.
This PR is an attempt to implement graceful shutdown for the API server process.

Partially fix https://github.com/argoproj/argo-cd/issues/18642

Signed-off-by: Leonardo Luz Almeida <leonardo_almeida@intuit.com>

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -560,8 +569,85 @@ func (a *ArgoCDServer) Run(ctx context.Context, listeners *Listeners) {
...
+
+		// Shutdown https server
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			err := httpsS.Shutdown(sCtx)
+			if err != nil {
+				log.Errorf("Error shutting down https server: %s", err)
+			}
+		}()
```
Comment (crenshaw-dev) [inline server/server.go:596]: Needs a nil check, since httpsS isn't necessarily set.

File: server/server.go [hunk 2]
```diff
@@ -560,8 +569,85 @@ func (a *ArgoCDServer) Run(ctx context.Context, listeners *Listeners) {
...
+				log.Errorf("Error shutting down metrics server: %s", err)
+			}
+		}()
+
+		// Shutdown tls server
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			tlsm.Close()
+		}()
```
Comment (crenshaw-dev) [inline server/server.go:620]: Needs a nil check, since tlsm isn't necessarily set.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Having trouble getting code coverage for api-server. I hoped graceful shutdown would help, so was testing this.

**Comment by andrii-korotkov-verkada** [comment]
I spoke with @leoluz about picking up this work. I've filed https://github.com/argoproj/argo-cd/pull/20981, since there's been merge conflicts and resolve conflicts didn't work perhaps due to age. I did some fixes and added co-authors.

**Pull request closed without merging**